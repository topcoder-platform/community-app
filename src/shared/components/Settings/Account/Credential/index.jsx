/**
 * Child component of Settings/Account/Credential renders the
 * 'Credential' section of account setting page.
 */
/* global document */
import _ from 'lodash';
import React from 'react';
import PT from 'prop-types';

import { PrimaryButton } from 'topcoder-react-ui-kit';

import Styles from './styles.scss';

export default class Credential extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      hasLength: false,
      hasLetter: false,
      hasSymbolNumber: false,
      passwordValid: false,
      showNewTips: false,
      focus: {
        'new-password-input': false,
        'current-password-input': false,
      },
    };

    this.onPasswordFocus = this.onPasswordFocus.bind(this);
    this.onPasswordBlur = this.onPasswordBlur.bind(this);
    this.onUpdatePassword = this.onUpdatePassword.bind(this);
    this.checkPassword = this.checkPassword.bind(this);
    this.toggleTypeAttribute = this.toggleTypeAttribute.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.profileState.updatingPassword
      && !nextProps.profileState.updatingPassword
      && !nextProps.settingsPageState.incorrectPassword
    ) {
      document.querySelector('#new-password-input').value = '';
      document.querySelector('#current-password-input').value = '';
      this.setState({ passwordValid: false });
    }
  }

  onUpdatePassword(e) {
    e.preventDefault();
    const newPassword = document.querySelector('#new-password-input').value;
    const currentPassword = document.querySelector('#current-password-input').value;
    this.props.updatePassword(
      this.props.profile, this.props.tokenV3,
      newPassword, currentPassword,
    );
  }

  onPasswordFocus(e) {
    const newState = { ...this.state };
    newState.focus[e.target.id] = true;

    if (e.target.id === 'new-password-input') {
      newState.showNewTips = true;
    }

    this.setState(newState);
  }

  onPasswordBlur(e) {
    const newState = { ...this.state };
    newState.focus[e.target.id] = false;

    if (e.target.id === 'new-password-input') {
      if (e.relatedTarget && e.relatedTarget.id === 'newPasswordCheckbox') {
        newState.showNewTips = true;
      } else {
        newState.showNewTips = false;
      }
    }

    this.setState(newState);
  }

  toggleTypeAttribute(inputId) {
    _.noop(this);
    const input = document.querySelector(`#${inputId}`);
    if (input.type === 'text') {
      input.type = 'password';
    } else {
      input.type = 'text';
    }
    input.focus();
  }

  checkPassword(e) {
    if (e.target.id === 'current-password-input') {
      this.props.clearIncorrectPassword();
    }

    const password = document.querySelector('#new-password-input').value;
    let hasLength = false;
    let hasLetter = false;
    let hasSymbolNumber = false;
    if (password) {
      hasLength = password.length >= 8;
      hasLetter = /[a-zA-Z]/.test(password);
      hasSymbolNumber = /[-!$@#%^&*()_+|~=`{}[\]:";'<>?,./]/.test(password) || /[\d]/.test(password);
    }
    const currentPassword = document.querySelector('#current-password-input').value;
    this.setState({
      ...this.state,
      hasLength,
      hasLetter,
      hasSymbolNumber,
      passwordValid: currentPassword.length && hasLength && hasLetter && hasSymbolNumber,
    });
  }

  render() {
    const {
      profile,
      profileState,
      settingsPageState,
    } = this.props;

    const { updatingPassword } = profileState;
    const { incorrectPassword } = settingsPageState;

    return (
      <div className="settings-section" styleName="credentials">
        <div className="section-info">
          <h2>Credentials</h2>
          <div className="description">Used to log in to your account and cannot be edited. Please contact support@topcoder.com if you need to make changes.</div>
        </div>
        <div className="section-fields" styleName="section-fields">
          <div className="form-label username">Username</div>
          <input name="username" value={profile.handle} disabled className="form-field grey" />
          <div className="form-label">Email</div>
          <input name="email" value={profile.email} disabled className="form-field grey" />
          {
            profileState.credential && profileState.credential.hasPassword &&
            <div>
              <form name="newPasswordForm" noValidate>
                <div className="form-label">Current password</div>
                <div styleName="validation-bar" className="form-field">
                  <div styleName={`password toggle-password ${this.state.focus['current-password-input'] ? 'focus' : ''}`}>
                    <input id="current-password-input" styleName="password-input" onChange={this.checkPassword} onFocus={this.onPasswordFocus} onBlur={this.onPasswordBlur} name="currentPassword" type="password" placeholder="Password" required />
                    <label htmlFor="currentPasswordCheckbox">
                      <input type="checkbox" id="currentPasswordCheckbox" onChange={() => this.toggleTypeAttribute('current-password-input')} />
                      Show
                    </label>
                  </div>
                  {
                    incorrectPassword &&
                    <div className="form-input-error">
                      <p>
                        Your current password is incorrect.
                        Please check that you entered the right one.
                      </p>
                    </div>
                  }
                </div>
                <div className="form-label">New Password</div>
                <div styleName="validation-bar" className="form-field">
                  <div styleName={`password toggle-password ${this.state.focus['new-password-input'] ? 'focus' : ''}`}>
                    <input id="new-password-input" styleName="password-input" onChange={this.checkPassword} onFocus={this.onPasswordFocus} onBlur={this.onPasswordBlur} name="password" type="password" placeholder="Create new password" minLength="8" maxLength="64" required />
                    <label htmlFor="newPasswordCheckbox">
                      <input type="checkbox" id="newPasswordCheckbox" onChange={() => this.toggleTypeAttribute('new-password-input')} />
                      Show
                    </label>
                  </div>
                  <div id="password-tips" styleName="tips password-tips" className={this.state.showNewTips ? '' : 'hidden'}>
                    <div styleName="arrow" />
                    <h3>Your password must have:</h3>
                    <p styleName={this.state.hasLength ? 'has-length-between-range' : ''}>At least 8 characters</p>
                    <p styleName={this.state.hasLetter ? 'has-letter' : ''}>At least one letter</p>
                    <p styleName={this.state.hasSymbolNumber ? 'has-symbol-or-number' : ''}>At least one number or symbol</p>
                  </div>
                </div>
                <div styleName="button-container">
                  <PrimaryButton
                    disabled={!this.state.passwordValid || updatingPassword}
                    onClick={this.onUpdatePassword}
                    theme={{ button: Styles['save-password-button'] }}
                  >
                    {
                      !updatingPassword && 'Change Password'
                    }
                    {
                      updatingPassword && <i className="fa fa-spinner fa-spin" />
                    }
                  </PrimaryButton>
                </div>
              </form>
            </div>
          }
          {
            profileState.credential && profileState.credential.hasPassword === false &&
            <div>
              <p>
                You joined Topcoder by using an external account,
                so we don&quot;t have a password for you.
              </p>
            </div>
          }
        </div>
      </div>
    );
  }
}

Credential.propTypes = {
  tokenV3: PT.string.isRequired,
  profile: PT.shape().isRequired,
  profileState: PT.shape().isRequired,
  settingsPageState: PT.shape().isRequired,
  updatePassword: PT.func.isRequired,
  clearIncorrectPassword: PT.func.isRequired,
};
