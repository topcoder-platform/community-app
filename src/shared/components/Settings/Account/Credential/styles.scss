@import "~styles/mixins";

.credentials {
  .section-fields {
    display: flex;
    flex-direction: column;

    .validation-bar {
      width: 100%;
      display: block;
      position: relative;

      & input[type='text'],
      & input[type='password'] {
        width: 100% !important;
      }

      .password {
        @include merriweather-sans-regular;

        padding-left: 0;
        height: 40px;
        width: 100%;
        font-size: 15px;
        line-height: 20px;
        color: #3d3d3d;
        margin-bottom: 10px;
        outline: none;
        background: #fff;
        border: 1px solid #b7b7b7;
        box-shadow: inset 0 0 2px 0 hsla(0, 0%, 72%, 0.2);
        border-radius: 2px;
        cursor: auto;
        transition: all 0.15s;

        &:hover {
          cursor: text;
        }

        input,
        input:hover,
        input:focus {
          background: none;
          border: none;
          box-shadow: none;
        }
      }
    }

    .button-container {
      display: flex;
      flex-direction: row;
      width: 100%;
    }
  }

  // Toggle password directive
  .toggle-password {
    @include merriweather-sans-regular;

    height: 40px;
    width: 100%;
    font-size: 15px;
    line-height: 20px;
    margin-bottom: 10px;
    background: #fff;
    border-radius: 2px;
    cursor: auto;
    transition: all 0.15s;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    position: relative;
    padding: 0;

    &.focus {
      border: 1px solid #0096ff !important;
    }

    input.password-input {
      margin-bottom: 0;
      outline: 0;
      width: 100%;
      padding-left: 10px;
      padding-right: 64px;
      height: 100%;
      border: none;

      @include placeholder {
        @include sofia-pro-regular;

        color: $tc-gray-40;
        font-size: 12px;
        text-transform: uppercase;
        transition: opacity 0.15s;
      }
    }

    label {
      @include sofia-pro-medium;

      display: flex !important;
      line-height: 20px !important;
      position: absolute;
      top: 10px;
      right: 0;
      text-transform: uppercase;
      font-size: 10px;
      color: #a3a3ae;
      height: 20px;
      width: 60px;
      padding: 0;

      input[type=checkbox] {
        height: 14px;
        width: 14px;
        visibility: visible;
      }
    }
  }
  // FORM TIPS
  // ----------------------------------------------------------------------
  // probably structure needs to be redone

  $tips-shadow: 0 1px 6px 1px rgba(0, 0, 0, 0.2);
  $tips-background: $tc-gray-neutral-light;

  // Tip conatiner
  .tips {
    background-color: $tips-background;
    border-radius: 4px;
    opacity: 1;
    padding: 20px;
    padding-bottom: 20px;
    transition: all 0.1s ease-in-out;
    z-index: 1000;

    @media only screen and (min-width: 768px) {
      left: 105%;
    }

    @media screen and (min-width: 915px) {
      box-shadow: $tips-shadow;
      width: 320px;
      padding-bottom: 10px;
      position: absolute;
      text-align: left;
      top: -10px;
    }
    // The tip container is hidden
    &.ng-hide {
      opacity: 0;
    }
    // Tip container arrow element
    .arrow {
      display: none;

      @media screen and (min-width: 915px) {
        display: block;
        height: 34px;
        width: 24px;
        left: -24px;
        overflow: hidden;
        position: absolute;
        top: 10px;

        &::after {
          transform: rotate(45deg);
          -moz-transform: rotate(45deg);
          -ms-transform: rotate(45deg);
          -o-transform: rotate(45deg);
          -webkit-transform: rotate(45deg);
          background-color: $tips-background;
          box-shadow: $tips-shadow;
          content: '';
          display: block;
          height: 12px;
          width: 12px;
          left: 18px;
          position: absolute;
          top: 10px;
        }
      }
    }

    h3 {
      @include sofia-pro-medium;

      font-size: 14px;
      padding-bottom: 15px;

      @media only screen and (max-width: 915px) {
        font-size: 16px;
      }
    }

    p {
      @include merriweather-sans-regular;

      color: $tc-gray-90;
      font-size: 13px;
      line-height: 1.3;
      padding-bottom: 10px;
      position: relative;
      text-align: left;
    }

    p::before {
      @include sofia-pro-medium;

      content: '';
      width: 20px;
      font-size: 14px;
      left: 0;
      position: absolute;
      top: 0;
    }
  }

  // Password tips show checkmark validation
  // TODO: refactor name to be context independent, like .validation-tips
  .password-tips {
    p {
      padding-left: 20px;
    }

    p::before {
      color: $tc-green;
      content: '\2713 ';
      display: none;
    }

    .has-length-between-range::before,
    .has-letter::before,
    .has-symbol-or-number::before,
    .has-number::before {
      display: block;
    }
  }

  .save-password-button {
    @include sofia-pro-medium;

    margin-left: 0;
    height: 40px;
    text-transform: uppercase;
    font-size: 12px;
  }
}

@media only screen and (min-width: 768px) {
  .section-fields {
    margin-top: 0;
    width: 100%;
  }
}
