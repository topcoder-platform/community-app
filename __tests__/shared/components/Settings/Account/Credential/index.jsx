import React from 'react';
import Renderer from 'react-test-renderer/shallow';

import Credential from 'components/Settings/Account/Credential';

import userProfile from '../../__mocks__/user-profile.json';
import profileState from '../../__mocks__/profile-state.json';

const rnd = new Renderer();

it('renders credential section of account setting page correctly', () => {
  rnd.render((<Credential
    tokenV3=""
    profile={userProfile}
    profileState={profileState}
    settingsPageState={{}}
    updatePassword={() => {}}
    clearIncorrectPassword={() => {}}
  />));
  expect(rnd.getRenderOutput()).toMatchSnapshot();
});
