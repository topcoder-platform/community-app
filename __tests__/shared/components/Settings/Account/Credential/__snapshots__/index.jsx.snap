// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders credential section of account setting page correctly 1`] = `
<div
  className="settings-section src-shared-components-Settings-Account-Credential-___styles__credentials___1FM_Q"
>
  <div
    className="section-info"
  >
    <h2>
      Credentials
    </h2>
    <div
      className="description"
    >
      Used to log in to your account and cannot be edited. Please contact support@topcoder.com if you need to make changes.
    </div>
  </div>
  <div
    className="section-fields src-shared-components-Settings-Account-Credential-___styles__section-fields___3lqw4"
  >
    <div
      className="form-label username"
    >
      Username
    </div>
    <input
      className="form-field grey"
      disabled={true}
      name="username"
      value="tcscoder"
    />
    <div
      className="form-label"
    >
      Email
    </div>
    <input
      className="form-field grey"
      disabled={true}
      name="email"
      value="tcscoder@gmail.com"
    />
    <div>
      <form
        name="newPasswordForm"
        noValidate={true}
      >
        <div
          className="form-label"
        >
          Current password
        </div>
        <div
          className="form-field src-shared-components-Settings-Account-Credential-___styles__validation-bar___20izI"
        >
          <div
            className="src-shared-components-Settings-Account-Credential-___styles__password___qv318 src-shared-components-Settings-Account-Credential-___styles__toggle-password___38ak9"
          >
            <input
              className="src-shared-components-Settings-Account-Credential-___styles__password-input___G4PX3"
              id="current-password-input"
              name="currentPassword"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              required={true}
              type="password"
            />
            <label
              htmlFor="currentPasswordCheckbox"
            >
              <input
                id="currentPasswordCheckbox"
                onChange={[Function]}
                type="checkbox"
              />
              Show
            </label>
          </div>
        </div>
        <div
          className="form-label"
        >
          New Password
        </div>
        <div
          className="form-field src-shared-components-Settings-Account-Credential-___styles__validation-bar___20izI"
        >
          <div
            className="src-shared-components-Settings-Account-Credential-___styles__password___qv318 src-shared-components-Settings-Account-Credential-___styles__toggle-password___38ak9"
          >
            <input
              className="src-shared-components-Settings-Account-Credential-___styles__password-input___G4PX3"
              id="new-password-input"
              maxLength="64"
              minLength="8"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Create new password"
              required={true}
              type="password"
            />
            <label
              htmlFor="newPasswordCheckbox"
            >
              <input
                id="newPasswordCheckbox"
                onChange={[Function]}
                type="checkbox"
              />
              Show
            </label>
          </div>
          <div
            className="hidden src-shared-components-Settings-Account-Credential-___styles__tips___KL4iB src-shared-components-Settings-Account-Credential-___styles__password-tips___CHTar"
            id="password-tips"
          >
            <div
              className="src-shared-components-Settings-Account-Credential-___styles__arrow___3Vl7Y"
            />
            <h3>
              Your password must have:
            </h3>
            <p
              className=""
            >
              At least 8 characters
            </p>
            <p
              className=""
            >
              At least one letter
            </p>
            <p
              className=""
            >
              At least one number or symbol
            </p>
          </div>
        </div>
        <div
          className="src-shared-components-Settings-Account-Credential-___styles__button-container___34PHE"
        >
          <ThemedButton
            active={false}
            composeAdhocTheme="deeply"
            composeContextTheme="softly"
            disabled={true}
            enforceA={false}
            mapThemrProps={[Function]}
            onClick={[Function]}
            onMouseDown={null}
            openNewTab={false}
            replace={false}
            size={null}
            theme={
              Object {
                "button": "src-shared-components-Settings-Account-Credential-styles___save-password-button___1p4lgx",
              }
            }
            themePriority="adhoc-context-default"
            to={null}
          >
            Change Password
          </ThemedButton>
        </div>
      </form>
    </div>
  </div>
</div>
`;
